### ISSUE TITLE
[BUG]: Link Write Rule causes Recursion Error

### ISSUE THREAD

**Issue by mike-pisman** [roman-right/beanie#675]
**Describe the bug**

After upgrading dependencies, including from  Pydantic 1 to Pydantic 2, saving  files with Link `Write` rule causes RecursionError

`RecursionError: maximum recursion depth exceeded while calling a Python object`

**To Reproduce**
```python
await MyDocument.save(new_document, link_rule=WriteRules.WRITE)
```

or

```python
 await new_document.save(link_rule=WriteRules.WRITE)
```

**Expected behavior**
It used to work like a charm and create new documents when specifying links.

**Additional context**
There are no errors, if I save documents without writing links, i.e.

```python
await MyDocument.save(new_document)
```

However, the linked documents do not get written to MongoDB.

I am not sure if the new version of pydantic is causing the issue, so would be nice, if someone can confirm.

Pydantic version is Version: 2.3.0
Beanie version is 1.21.0

### Error observed (Error log)

You can see in the log bellow,
that the WRITE rule causes the interpreter to go to `beanie/odm/actions.py` line 219

```python
 result = await f(self, *args, skip_actions=skip_actions, **kwargs)
```

then to `beanie/odm/utils/state.py` line 66

```python
result = await f(self, *args, **kwargs)
```

then to `beanie/odm/documents.py` line 526

```python
await obj.save(
```

then again to `beanie/odm/actions.py` line 219

```python
 result = await f(self, *args, skip_actions=skip_actions, **kwargs)
```

and again to `beanie/odm/utils/state.py` line 66

```python
result = await f(self, *args, **kwargs)
```

finally to `beanie/odm/documents.py`  line 514

```python
await value.save(
```

and then this loop repeats itself until maximum depth of recursion is reached.

<details>
  <summary>Error Log</summary>

```Console
await Resource.save(self, link_rule=WriteRules.WRITE)
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/actions.py", line 219, in wrapper
    result = await f(self, *args, skip_actions=skip_actions, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/utils/state.py", line 66, in wrapper
    result = await f(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/documents.py", line 526, in save
    await obj.save(
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/actions.py", line 219, in wrapper
    result = await f(self, *args, skip_actions=skip_actions, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/utils/state.py", line 66, in wrapper
    result = await f(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/documents.py", line 514, in save
    await value.save(
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/actions.py", line 219, in wrapper
    result = await f(self, *args, skip_actions=skip_actions, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/utils/state.py", line 66, in wrapper
    result = await f(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/documents.py", line 526, in save
    await obj.save(
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/actions.py", line 219, in wrapper
    result = await f(self, *args, skip_actions=skip_actions, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/truenas/home/mpisman/Desktop/unipoll/API/venv/lib/python3.11/site-packages/beanie/odm/utils/state.py", line 66, in wrapper
    result = await f(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
</details>

I hope this helps.

**Comment by roman-right** [comment]
Hi @mike-pisman ,
Thank you for the report.
Please, provide one module example to me be able to reproduce this.

**Comment by mike-pisman** [comment]
@roman-right, thank you for the reply, and sorry for the wait. 
### TLDR: 
The error was my fault, which can be probably solved by using Backlinks(last time I tried backlinks were still in development). The weird part that with *Pydantic 1.10.9* and *Beanie 1.19.9* I do **not** see that error, and my code works fine. 

### Error cause
I had a chance to test a few things. Beanie seems to work fine, it was an error on my end.

Basically, I was creating two **Documents** and adding second document into the first one, while the second document had a link to the first document. It looks something like:
```python
 
class DocumentA(Document):
    childDocs: list[Link[DocumentB]]

class DocumentB(Document):
    parentDoc: Link[DocumentA]
```

And when creating documents I would create an infinite loop of linking...
```python
newParentDoc = await DocumentA().create()

newChildDoc = await DocumentB(parentDoc=newParentDoc)

newParentDoc.childDocs.append(newChildDoc)

await newParentDoc.save(link_rule=WriteRules.WRITE)
```

### Thank you again for the reply and creating this amazing library!
